pkg/kubelet/cm/hugepagehandler/hugepage_handler.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hugepagehandler
+
+import (
+	"strings"
+
+	"k8s.io/api/core/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+
+	units "github.com/docker/go-units"
+	cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
+	libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
+
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+type runtimeService interface {
+	UpdateContainerResources(id string, resources *runtimeapi.LinuxContainerResources) error
+}
+
+// Handler interface provides methods for Kubelet to set container hugetlb cgroup. 
+type Handler interface {
+	// Start is called during Kubelet initialization.
+	Start(containerRuntime runtimeService)
+
+	// AddContainer is called between container create and container start
+	// so that initial hugepage limit settings can be written through to the
+	// container runtime before the first process begins to execute.
+	AddContainer(p *v1.Pod, c *v1.Container, containerID string) error
+}
+
+type handler struct {
+	// containerRuntime is the container runtime service interface needed
+	// to make UpdateContainerResources() calls against the containers.
+	containerRuntime runtimeService
+}
+
+var _ Handler = &handler{}
+
+// NewHandler creates new handler
+func NewHandler() (Handler, error) {
+	handler := &handler{}
+	return handler, nil
+}
+
+func (m *handler) Start(containerRuntime runtimeService) {
+	m.containerRuntime = containerRuntime
+}
+
+func (m *handler) AddContainer(p *v1.Pod, c *v1.Container, containerID string) error {
+
+	hugepageLimits := []*runtimeapi.HugepageLimit{}
+	pageSizes := sets.NewString()
+
+	for resourceObj, amountObj := range c.Resources.Limits {
+		resourceStr := string(resourceObj)
+
+		//Check if resourceStr is "hugepages-<pageSize>", api.ResourceHugePagesPrefix = "hugepages-"
+		if !strings.HasPrefix(resourceStr, api.ResourceHugePagesPrefix) {
+			continue
+		} else {
+			hugepageSizeStr := strings.TrimPrefix(resourceStr, api.ResourceHugePagesPrefix)
```

Comment:
And also, is the plan to still write to the pod-level cgroup with the sum of the requested huge page memory? If we eventually want to remove that, I think it is ok to keep the logic here.

Reply:
> A lot of this logic is now duplicated https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cgroup_manager_linux.go#L386-L404. Do you think it is a good idea to separate the parsing code into a separate place?

Maybe ```k8s.io/kubernetes/pkg/apis/core/v1/helper``` ? I have no idea yet.

Reply:
> Maybe k8s.io/kubernetes/pkg/apis/core/v1/helper ? I have no idea yet.

Not sure myself. Maybe @mattjmcnaughton have some thoughts

Reply:
In the case of conventional memory, kubelet also sets merged memory limits on pod level cgroup.
Individual container level cgroup also has own limit for memory.
Maybe these duplicated limits(on both of pod and container level cgroup) are not intended.
Because at the early version of Kubernetes there were no CRI and OCI spec.

`And also, is the plan to still write to the pod-level cgroup with the sum of the requested huge page memory`

I think it depends on the issue, which is mentioned by @derekwaynecarr , which is `hugetlb miss-charging issues between pod and container level cgroup`.

If the only way to prevent the miss-charging issue is removing merged limits from the pod level cgroup.
I would content remove it.

Reply:
@odinuge Hi, i'm hyunsung, working with @bg-chun , @ohsewon.
I agreed with you, but it can be hard to combine that into a function gracefully in my idea.
Because the types(`c.Resources.Limits`, `resourceConfig.HugePageLimit`) are tricky to match.

Reply:
Ahh, I see. I prefer as much reuse as possible, but somethings it is unavoidable. As long as there is good test coverage, it should be ok. I think we atleast should reuse the line

```go
units.CustomSize("%g%s", float64(pageSize), 1024.0, libcontainercgroups.HugePageSizeUnitList)
```
so that we can test (and therefore document) it properly, to avoid regressions.